### PR TITLE
Enhance variable group key vault variable configuration

### PR DESCRIPTION
**Pull request by xuzhang3**
## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [ ] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

Issue Number: #252 
1. Variable group with key  vault will verify if the variables exists in AzureRM key vault.
2. Variable group key vault need workable AzureRM service connection and existed key vault secrets,  disable the acctest `TestAccVariableGroupKeyVault_CreateAndUpdate`.
Manual test result:
image
image
image
image

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [x] No

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Any relevant logs, error output, etc?

(If it’s long, please paste to https://ghostbin.com/ and insert the link here.)

### CODE REVIEW

File: azuredevops/internal/service/taskagent/resource_variable_group.go [hunk 1]
```diff
@@ -338,24 +361,95 @@ func expandVariableGroupParameters(d *schema.ResourceData) (*taskagent.VariableG
...
-			Vault:             converter.String(keyVaultValues[vgName].(string)),
+			ServiceEndpointId: &serviceEndpointUUID,
+			Vault:             &kvName,
 		}
 
 		variableGroup.Type = converter.String(azureKeyVaultType)
-	}
+		if azureKVSecrets, err := azureKVSecrets(clients, *projectID, kvName, serviceEndpointID); err != nil {
+			return nil, nil, err
+		} else {
```
Comment (EliiseS) [inline azuredevops/internal/service/taskagent/resource_variable_group.go]: This else is redundant due to the return statement in the if. You can just remove the else nesting.

File: azuredevops/internal/service/taskagent/resource_variable_group.go [hunk 2]
```diff
@@ -338,24 +361,95 @@ func expandVariableGroupParameters(d *schema.ResourceData) (*taskagent.VariableG
...
+						"KeyVaultName": kvName,
+					},
+				},
+				ResultTransformationDetails: &serviceendpoint.ResultTransformationDetails{},
+			},
+			Project:    &projectID,
+			EndpointId: &serviceEndpointID,
+		})
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get the Azure Key valut secrets. %v ", err)
```
Comment (EliiseS) [inline azuredevops/internal/service/taskagent/resource_variable_group.go]:
```suggestion
		return nil, fmt.Errorf("Failed to get the Azure Key vault secrets. %v ", err)
```

File: azuredevops/internal/service/taskagent/resource_variable_group.go [hunk 3]
```diff
@@ -36,9 +40,28 @@ const (
 	azureKeyVaultType                         = "AzureKeyVault"
 	invalidVariableGroupIDErrorMessageFormat  = "Error parsing the variable group ID from the Terraform resource data: %v"
 	flatteningVariableGroupErrorMessageFormat = "Error flattening variable group: %v"
-	expandingVariableGroupErrorMessageFormat  = "Error expanding variable group resource data: %+v"
+	expandingVariableGroupErrorMessageFormat  = "Expanding variable group resource data: %+v"
```
Comment (EliiseS) [inline azuredevops/internal/service/taskagent/resource_variable_group.go:43]: Why remove the word error here? Why remove it for this error but not the others?

File: azuredevops/internal/service/taskagent/resource_variable_group.go [hunk 4]
```diff
@@ -338,24 +361,95 @@ func expandVariableGroupParameters(d *schema.ResourceData) (*taskagent.VariableG
...
+					kvName)
+			} else {
+				variableGroup.Variables = &kvVariables
+			}
+		}
+	}
 	return variableGroup, projectID, nil
 }
 
+func azureKVSecrets(clients *client.AggregatedClient, projectID string, kvName string, serviceEndpointID string) (azureKVSecrets map[string]taskagent.AzureKeyVaultVariableValue, error error) {
```
Comment (EliiseS) [inline azuredevops/internal/service/taskagent/resource_variable_group.go]: To describe this function better rename to getAzureKVSecrets

File: azuredevops/internal/service/taskagent/resource_variable_group.go [hunk 5]
```diff
@@ -554,3 +648,11 @@ func flattenAllowAccess(d *schema.ResourceData, definitionResource *[]build.Defi
 	}
 	d.Set(vgAllowAccess, allowAccess)
 }
+
+func getSecretName(secretID string) (secret string) {
+	if len(secretID) == 0 {
+		return ""
+	}
+	secretURL := strings.Split(secretID, "/")
+	return secretURL[len(secretURL)-1]
```
Comment (EliiseS) [inline azuredevops/internal/service/taskagent/resource_variable_group.go:658]: I'm a bit confused, what's happening here?

File: azuredevops/internal/service/taskagent/resource_variable_group.go [hunk 6]
```diff
@@ -338,24 +361,95 @@ func expandVariableGroupParameters(d *schema.ResourceData) (*taskagent.VariableG
...
+				kvSecretName := variable.(map[string]interface{})[vgName].(string)
+				if kv, ok := azureKVSecrets[kvSecretName]; ok {
+					kvVariables[kvSecretName] = kv
+				} else {
+					invalidVariables = append(invalidVariables, kvSecretName)
+				}
+			}
 
+			if len(invalidVariables) > 0 {
+				return nil, nil, fmt.Errorf("Invaild  Key Vault variables : ( %s ) , can not find in Azure Key Vault: ( %s ) ",
```
Comment (EliiseS) [inline azuredevops/internal/service/taskagent/resource_variable_group.go]:
```suggestion
				return nil, nil, fmt.Errorf("Invalid Key Vault variables: ( %s ) , can not find in Azure Key Vault: ( %s ) ",
```

File: azuredevops/internal/acceptancetests/resource_variable_group_test.go [hunk 7]
```diff
@@ -67,6 +67,7 @@ func TestAccVariableGroup_CreateAndUpdate(t *testing.T) {
 }
 
 func TestAccVariableGroupKeyVault_CreateAndUpdate(t *testing.T) {
+	t.Skip("Skipping test TestAccVariableGroupKeyVault_CreateAndUpdate: test resource limit")
```
Comment (EliiseS) [inline azuredevops/internal/acceptancetests/resource_variable_group_test.go]:
```suggestion
	t.Skip("Skipping test TestAccVariableGroupKeyVault_CreateAndUpdate: azure key vault not provisioned on test infrastructure")
```

**Review by EliiseS — CHANGES_REQUESTED** [review]

File: azuredevops/internal/service/taskagent/resource_variable_group.go (hunk 3 above)
Comment (xuzhang3) [reply azuredevops/internal/service/taskagent/resource_variable_group.go:43]: Terraform will automatically add the `Error:` prefix to each error message.
image

File: azuredevops/internal/service/taskagent/resource_variable_group.go (hunk 5 above)
Comment (xuzhang3) [reply azuredevops/internal/service/taskagent/resource_variable_group.go:658]: Azure key secret ID returned in http format. The real secret name is at the end of the ID.
Service response:
```json
{
            "contentType": "1222",
            "id": "https://xzadotest.vault.azure.net/secrets/secret1",
            "attributes": {
                "enabled": true,
                "exp": 1675424159,
                "created": 1611734011,
                "updated": 1612353657,
                "recoveryLevel": "Recoverable+Purgeable"
            },
            "tags": {}
        },
```

File: azuredevops/internal/service/taskagent/resource_variable_group.go (hunk 3 above)
Comment (EliiseS) [reply azuredevops/internal/service/taskagent/resource_variable_group.go:43]: Should we remove it from the other ones too then?

**Review by EliiseS — APPROVED** [review]

File: azuredevops/internal/service/taskagent/resource_variable_group.go (hunk 3 above)
Comment (xuzhang3) [reply azuredevops/internal/service/taskagent/resource_variable_group.go:43]: I'm tend to remove it.

**Pull request merged**